server/src/main/java/org/opensearch/tasks/TaskCancellation.java

```diff
@@ -42,10 +42,25 @@ public List<Reason> getReasons() {
         return reasons;
     }
 
+    public List<Runnable> getOnCancelCallbacks() {
+        return onCancelCallbacks;
+    }
+
     public String getReasonString() {
         return reasons.stream().map(Reason::getMessage).collect(Collectors.joining(", "));
     }
 
+    public TaskCancellation merge(final TaskCancellation other) {
+        if (other == this) {
+            return this;
+        }
+        final List<Reason> newReasons = new ArrayList<>(reasons);
+        newReasons.addAll(other.getReasons());
+        final List<Runnable> newOnCancelCallbacks = new ArrayList<>(onCancelCallbacks);
+        newOnCancelCallbacks.addAll(other.onCancelCallbacks);
+        return new TaskCancellation(task, newReasons, newOnCancelCallbacks);
```

Comment:
Callbacks are mainly the counters kept by `TaskResourceUsageTracker` and `SearchBackpressureState`. For duplicate tasks it seems like we will now end up incrementing the counters twice ?

Reply:
No it will not double count but rather count at two levels separately i,e
`SearchBackpressureState` keeps the stats at `SearchTask` level and `SearchShardTask` level while 
`TaskResourceUsageTracker` keeps the stats at individual tracker(heap|cpu|elapsedTime) level.
The relation here is 
```
SearchBackpressureState(cancellationCount) = Sum(cancellationCount in resourceTrackerY for resourceTrackerY in allResourceTrackers)
```

for more clarification, **We are sending the searchBackpressureState taskType cancellation callback for all the trackers for the taskType
```
taskResourceUsageTracker.ifPresent(
                tracker -> taskCancellations.addAll(
                    tracker.getTaskCancellations(cancellableTasks.get(taskType), searchBackpressureStates.get(taskType)::incrementCancellationCount)
                )
            );
```

Reply:
Thanks for explaining this. Probably this is the existing behavior but we may end up counting the cancellation twice but for different reasons. For example: if same task is eligible for cancellations both by Heap and CPU usage, we will count cancellation count as 2 for `SearchBackpressureState` whereas only 1 task is getting cancelled. Depending on how these stats are used it may not reflect the correct behavior. I think at the minm. we should update the [documentation](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/search-backpressure/) to explain the stats correctly.

Reply:
Ah! I see this will double count. Good Catch I will address this.